sympy/polys/matrices/ddm.py

```diff
@@ -331,3 +379,9 @@ def charpoly(a):
 
 
 from .sdm import SDM
+
+def ddm_applyfunc(M, f, domain):
+    return [[domain.convert(f(x)) for x in Mi] for Mi in M]
```

Comment:
It is generally more efficient to use `convert_from` rather than `convert`. Also mostly the old and new domains will be the same in which case we can avoid converting altogether.

Actually though this is misunderstanding the purpose of the domain argument which is that it should specify the domain in the case that `f` will return elements of a different domain. The point is not that applyfunc should call convert but that in some cases `f` will already have converted the elements to another domain in which case the caller of `pplyfunc` needs to tell `DomainMatrix` what that other domain is.

Reply:
So you want me to use `convert` or `convert_from` only when the output `domain` is specified?

Reply:
There is no need to use convert at all. The domain argument is needed if the function `f` already returns elements of a different domain.

Reply:
@oscarbenjamin what do you think about this?

```python
def sdm_applyfunc_nonzero(M, f, domain=None):
    Mf = {}
    for i, Mi in M.items():
        for j, Mij in Mi.items():
            fMij = f(M[i][j]) if domain is None else domain.convert(f(M[i][j]))
        
            if not domain.is_zero(fMij):
                try:
                    Mf[i][j] = fMij
                except KeyError:
                    Mf[i] = {j: fMij}

    return Mf
```

Do you think I should also pass current domain to functions `sdm_applyfunc` and `sdm_applyfunc_nonzero` and use `convert_from` instead of `convert` ?

Reply:
As I already said: there is no need to use convert at all.